File: C++/Sorting/heapSort.cpp
#include <bits/stdc++.h>
using namespace std;

void insert(vector<int> &heap, int value)
{
    heap.push_back(value);
    int i = heap.size() - 1;
    while( i>0)
    {
        int parent = (i-1)/2;
        if(heap[parent]<heap[i])
        {
            int temp = heap[i];
            heap[i] = heap[parent];
            heap[parent] = temp;
            i = parent;
        }
        else
        {
            break;
        }
    }

}

vector<int> makeHeap(int arr[], int n)
{
    vector<int> heap{};
    for(int i=0;i<n;i++)
    {
        insert(heap,arr[i]);
    }
    return heap;
} 

void heapify(vector<int> &heap, int size, int i)
{
    if(i>size)
        return;
    
    while(i<size)
    {
        int left = 2*i+1;
        int right = 2*i+2;
        int largest = i;
        if(left<size && heap[largest]<=heap[left])
            largest=left;
        if(right<size && heap[largest]<=heap[right])
            largest=right;
        if (largest!=i)
        {
            int temp = heap[i];
            heap[i] = heap[largest];
            heap[largest] = temp;
            i = largest;
        }
        else
            break;
    }
}

void del(vector<int> &heap, int index)
{
    if(index>=heap.size())
    {
        return;
    }

    heap[index] = heap[heap.size()-1];
    heap.pop_back();
    heapify(heap,heap.size(),index);

}

void print(vector<int>& heap)
{
    for(int i=0;i<heap.size();i++)
    {
        cout<<heap[i]<<" ";
    }
    cout<<endl;
}

void heapSort(vector<int> &heap)
{
    for(int i=heap.size()-1;i>0;i--)
    {
        int temp = heap[0];
        heap[0] = heap[i];
        heap[i] = temp;
        heapify(heap,i,0);
    }
}
int main()
{
    int arr[] = {1,1,-5,8,20,35,13,-11,3,4};
    vector<int> heap = makeHeap(arr,10);
    print(heap);
    heapSort(heap);
    print(heap);
}

File: Python/Sorting/heapSort.py
# max heap
def insert(heap:list, num:int)->list:
    heap.append(num)
    i = len(heap) - 1
    while i > 0:
        parent = (i - 1) // 2
        if heap[parent]<heap[i]:
            temp = heap[i]
            heap[i] = heap[parent]
            heap[parent] = temp
            i=parent
        else:
            break
    return heap

def heapify(arr:list)->list:
    heap = []
    for i in range(len(arr)):
        heap = insert(heap, arr[i])
    return heap


def delete(heap:list, value:int)->list:
    index=-1
    for i in range(len(heap)):
        if heap[i]==value:
            index=i
    if index==-1:
        raise f"Value {value} not found in the heap."
    
    heap[index] = heap[len(heap)-1]
    del heap[len(heap)-1]
    
    while index<len(heap)-1:    
        child = 2*index + 1
        
        if heap[child]>heap[index]:
            temp = heap[child]
            heap[child] = heap[index]
            heap[index] = temp
    
    return heap

def makeHeap(heap: list, i: int, n:int) -> list:
    if i >= n:
        return heap

    while i < n:
        child1 = 2 * i + 1
        child2 = 2 * i + 2
        largest = i

        if child1 < n and heap[child1] > heap[largest]:
            largest = child1
        if child2 < n and heap[child2] > heap[largest]:
            largest = child2

        if largest != i:
            heap[i], heap[largest] = heap[largest], heap[i]
            i = largest
        else:
            break

    return heap


def heapSort(heap:list)->list:
    heap = makeHeap(heap, len(heap)-1, len(heap))
    for i in range(len(heap)-1, 0, -1):
        temp = heap[0]
        heap[0] = heap[i]
        heap[i] = temp
        heap = makeHeap(heap, 0, i)

    return heap





heap=[]
arr = [1,2,3,4,5,6,7,8]
print("Array: ",arr)
heap = heapify(arr)
print("Heap:  ",heap)
heap = heapSort(heap)
print("Sorted:",heap)
